File: Python3/503. Next Greater Element II.py
class Solution:
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # length = len(nums) # 超时
        # if not length:
        #     return []
        # maxnum = max(nums)
        # nums *= 2
        # for i in range(length):
        #     if nums[i] < maxnum:
        #         for j in range(i+1, i+length):
        #             if nums[j] > nums[i]:
        #                 nums[i] = nums[j]
        #                 break
        #     else:
        #         nums[i] = -1
        # return nums[: length]
        
        length = len(nums) # stack 208ms
        if not length:
            return []
        nums *= 2
        stack = []
        for i in range(2*length-1, -1, -1):
            while stack and nums[i] >= stack[-1][1]:
                stack.pop()
            if stack:
                temp = stack[-1][1]
                stack.append((i, nums[i]))
                nums[i] = temp
            else:
                stack.append((i, nums[i]))
                nums[i] = -1
        return nums[: length]

File: Golang/503. Next Greater Element II.go
func nextGreaterElements(nums []int) []int { // 632ms
    length := len(nums)
    stack := make([]int, 0, length)
    ans := make([]int, length)
    for i := 2*length-1; i >= 0; i-- {
        for len(stack) > 0 && nums[i%length] >= stack[len(stack)-1] {
            stack = stack[: len(stack)-1]
        }
        if len(stack) > 0 {
            ans[i%length] = stack[len(stack)-1]
            stack = append(stack, nums[i%length])
        } else {
            ans[i%length] = -1
            stack = append(stack, nums[i%length])
        }
    }
    return ans
}